### PR TITLE
fix(mural): fallback to file backend when keyring is empty

### LINKED ISSUE #1942: fix: Mural CLI auto backend selects empty keyring instead of populated file credentials
### Component

Skills

## Bug Summary

When `MURAL_CREDENTIAL_BACKEND=auto`, the CLI can select keyring even when keyring has no Mural credentials. If file credentials are present, authentication should still succeed by falling back to the file backend.

## Observed Behavior

* `python -m mural auth status` fails with missing credential errors
* Behavior occurs on machines where keyring is installed but empty for the target Mural profile
* File credentials exist at `%APPDATA%\hve-core\mural.default.env`, but they are ignored in auto mode

## Impact

* Blocks first-time setup and recovery workflows
* Produces confusing auth failures even when valid credentials are already configured
* Increases support effort for environment setup

## Suggested Fix

Update backend resolution logic for `auto` mode:

1. Instantiate keyring backend
2. Check known credential keys for the current profile service
3. If keyring is empty, evaluate file backend contents
4. If file backend is populated, select file backend
5. Emit a one-time warning that keyring was available but empty and file backend was selected

## Affected Area

* `.github/skills/experimental/mural/scripts/mural/_backends.py`

## Expected Behavior

* In auto mode, backend selection should verify credential presence, not only backend availability
* If keyring is available but empty, CLI should fall back to file backend when file credentials exist
* Auth status and subsequent commands should proceed using valid file-backed credentials

## Environment

* OS: Windows
* Shell: PowerShell
* Python environment: `C:\GitHub\hve-core\.venv`
* Module path: `.github/skills/experimental/mural/scripts`

## Steps To Reproduce

1. Ensure keyring backend is available in Python environment
2. Ensure keyring has no values for Mural profile service keys
3. Create a valid credential file with client ID and client secret at `%APPDATA%\hve-core\mural.default.env`
4. Set backend mode to auto
5. Run:

```bash
python -m mural auth status
```

* Command fails due to unresolved credentials because empty keyring is selected first

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by akazlow**
# Pull Request

## Description

Fixed automatic credential backend resolution when the keyring is available but empty and the credential file contains usable values.

The file backend now remains available for immediate authentication, the warning explains that credentials are eligible for promotion during auth write flows, and promotion failure paths preserve complete credentials.

## Related Issue(s)

Fixes #1942

## Type of Change

Select all that apply:

**Code & Documentation:**

* [x] Bug fix (non-breaking change fixing an issue)
* [ ] New feature (non-breaking change adding functionality)
* [ ] Breaking change (fix or feature causing existing functionality to change)
* [ ] Documentation update

**Infrastructure & Configuration:**

* [ ] GitHub Actions workflow
* [ ] Linting configuration (markdown, PowerShell, etc.)
* [ ] Security configuration
* [ ] DevContainer configuration
* [ ] Dependency update

**AI Artifacts:**

* [ ] Reviewed contribution with `hve-builder` and addressed all actionable findings
* [ ] Copilot instructions (`.github/instructions/*.instructions.md`)
* [ ] Copilot prompt (`.github/prompts/*.prompt.md`)
* [ ] Copilot agent (`.github/agents/*.agent.md`)
* [ ] Copilot skill (`.github/skills/*/SKILL.md`)
* [ ] Copilot hook (`.github/hooks/*/*.json`)
* [ ] Eval spec added/updated for changed AI artifacts (`evals/`)

> Note for AI Artifact Contributors:
>
> * Agents: Research, indexing/referencing other project (using standard VS Code GitHub Copilot/MCP tools), planning, and general implementation agents likely already exist. Review `.github/agents/` before creating new ones.
> * Skills: Must include both bash and PowerShell scripts. See [Skills](../docs/contributing/skills.md).
> * Model Versions: Contributions **MUST** target models listed in the model catalog (`scripts/linting/model-catalog.json`) whose provider appears in `providerAllowlist` and whose status is `ga` or `preview`. Run `npm run lint:models` to validate references.
> * See [Agents Not Accepted](../docs/contributing/custom-agents.md#agents-not-accepted) and [Model Version Requirements](../docs/contributing/ai-artifacts-common.md#model-version-requirements).

**Other:**

* [x] Script/automation (`.ps1`, `.sh`, `.py`)
* [ ] Other (please describe):

## Sample Prompts (for AI Artifact Contributions)

Not applicable; no AI artifact definitions changed.

## Testing

* Passed `uv run pytest tests/test_credential_storage.py -q`: 27 passed, 20 Windows-specific skips.
* Passed the same focused suite with `CI=true`: 27 passed, 20 Windows-specific skips.
* Passed `uv run ruff check scripts tests`.
* Passed `uv run ruff format --check scripts tests` across 44 files.
* Passed `git diff --check`; only the two intended Mural Python files changed and no dependencies changed.
* `npm run validate:local` was attempted but exceeded the local execution window during repository-wide Markdown lint, so hosted CI remains required for the aggregate result.
* Manual testing was not performed.

### Required Checks

* [ ] Documentation is updated (N/A: no documentation files changed)
* [x] Files follow existing naming conventions
* [x] Changes are backwards compatible (if applicable)
* [x] Tests added for new functionality (if applicable)

### AI Artifact Contributions

* [ ] Used `hve-builder` review mode to review contribution
* [ ] Addressed all actionable findings from the `hve-builder` review
* [ ] Verified contribution follows common standards and type-specific requirements

### Required Local Checks

The following local-safe validation commands must pass before merging:

* [ ] Local validation aggregate: `npm run validate:local` (timed out during repository-wide Markdown lint; awaiting hosted CI)
* [ ] Documentation validation (N/A: no documentation files changed)
* [ ] Spell checking (covered by hosted CI)
* [ ] Link validation (N/A: no links changed)

## Security Considerations

* [x] This PR does not contain any sensitive or NDA information
* [ ] Any new dependencies have been reviewed for security issues (N/A: no dependencies changed)
* [ ] Security-related scripts follow the principle of least privilege (N/A: no security scripts changed)

## Additional Notes

* Auto mode remains unchanged when the keyring already contains credentials.
* Passive backend resolution does not write to the keyring; promotion remains limited to eligible auth write flows.
* Failed or unverifiable promotion preserves the credential file, and partial keyring writes are rolled back.

### CODE REVIEW

**Comment by akazlow** [comment]
## Review Summary

This fix addresses credential backend resolution when the keyring backend is selected but lacks usable secrets.

### Key Changes
- **_backends.py**: Added fallback logic to use file backend when keyring is unavailable/empty
- Prevents auth failures when credential selection doesn't match available backend state

### Impact
- Mural skill can now gracefully degrade to file-based auth instead of failing entirely
- Improves user experience in devcontainer/CI scenarios where keyring may not be initialized

### Verification Checklist
- [x] Backend resolution scope limited to single file
- [x] Fallback preserves original credential file path
- [x] No breaking changes to existing auth workflows

File: .github/skills/experimental/mural/scripts/mural/_backends.py [hunk 1]
```diff
@@ -230,6 +230,31 @@ def resolve_backend(profile: str = "default") -> CredentialBackend:
...
         try:
             selected = KeyringBackend()
+            service = _service_name_for(profile)
+            keyring_populated = False
+            for key in _KNOWN_CREDENTIAL_KEYS:
+                value = selected.get(service, key)
+                if value:
+                    keyring_populated = True
+                    break
+            if not keyring_populated:
```
Comment (katriendg) [inline .github/skills/experimental/mural/scripts/mural/_backends.py:241]: **Regression test (please address):** This new fallback path has no test. Could you add one to the `TestResolveBackend` class in `tests/test_credential_storage.py` asserting that, in `auto` mode, an available-but-empty keyring plus a populated credential file resolves to `FileBackend` and emits exactly one fallback warning (deduped per profile per process, mirroring `test_auto_fallback_warn_dedupes_per_profile_per_process`)? The existing `test_auto_returns_keyring_when_available` only covers the both-empty case, so this path is currently untested.

File: .github/skills/experimental/mural/scripts/mural/_backends.py [hunk 2]
```diff
@@ -230,6 +230,31 @@ def resolve_backend(profile: str = "default") -> CredentialBackend:
     elif selector == "auto":
         try:
             selected = KeyringBackend()
+            service = _service_name_for(profile)
```
Comment (katriendg) [inline .github/skills/experimental/mural/scripts/mural/_backends.py:239]: **Docstring update (please address):** The `resolve_backend` docstring above still describes auto fallback as happening only "when `_KeyringUnavailable` is raised." This PR adds a second trigger (keyring available but empty → file backend when the file is populated). Please extend the docstring to document the new path and its one-shot warning so the contract stays accurate.

File: .github/skills/experimental/mural/scripts/mural/_backends.py [hunk 3]
```diff
@@ -230,6 +230,31 @@ def resolve_backend(profile: str = "default") -> CredentialBackend:
...
+            keyring_populated = False
+            for key in _KNOWN_CREDENTIAL_KEYS:
+                value = selected.get(service, key)
+                if value:
+                    keyring_populated = True
+                    break
+            if not keyring_populated:
+                file_backend = FileBackend(file_path)
+                if file_path.exists():
+                    _check_credential_file_perms(file_path, os.environ)
```
Comment (katriendg) [inline .github/skills/experimental/mural/scripts/mural/_backends.py]: **Perms-check consistency (your call, non-blocking):** This path runs `_check_credential_file_perms()` at resolve time, which can raise `MuralError` on POSIX for a bad mode/owner. The keyring-*unavailable* fallback just below constructs `FileBackend(file_path)` without a resolve-time perms check. On Windows this is a no-op, but on POSIX the empty-keyring fallback is now stricter than the unavailable-keyring one. Worth deciding whether to apply the check to both paths or neither, for symmetry.

File: .github/skills/experimental/mural/scripts/mural/_backends.py [hunk 4]
```diff
@@ -230,6 +230,31 @@ def resolve_backend(profile: str = "default") -> CredentialBackend:
...
+                file_entries = file_backend._read_all()
+                file_populated = any(
```
Comment (katriendg) [inline .github/skills/experimental/mural/scripts/mural/_backends.py]: **Optional refactor:** These keyring / file "is populated" probes duplicate logic already in `_maybe_warn_concurrent_state`. A small shared helper (e.g. `_backend_has_credentials(backend, service)`) would centralize it and reduce drift risk if the key set or probing logic changes later. Purely optional.

File: .github/skills/experimental/mural/scripts/mural/_backends.py [hunk 5]
```diff
@@ -230,6 +230,31 @@ def resolve_backend(profile: str = "default") -> CredentialBackend:
     elif selector == "auto":
         try:
             selected = KeyringBackend()
+            service = _service_name_for(profile)
+            keyring_populated = False
+            for key in _KNOWN_CREDENTIAL_KEYS:
+                value = selected.get(service, key)
```
Comment (katriendg) [inline .github/skills/experimental/mural/scripts/mural/_backends.py]: **Info, no action needed:** Because this `.get()` runs inside the outer `try`, a keyring read error *after* successful construction is caught by `except _KeyringUnavailable` and reported as "unavailable… falling back" rather than "empty." Reasonable behavior — just flagging that the message can read slightly off in that edge case.

**Review by katriendg — COMMENTED** [review]
Thanks for filing both the issue (#1942) and this fix, @akazlow — clear, well-scoped report and a tidy, minimal change. I validated it locally: an available-but-empty keyring plus a populated credential file now resolves to `FileBackend` and emits the one-time fallback warning, matching all five steps of your suggested fix. The existing `test_credential_storage.py` suite (38 tests) still passes and `ruff` is clean. 🎉

Two process asks before we go further:

- Please update the PR description to follow the repo's [pull request template](https://github.com/microsoft/hve-core/blob/main/.github/PULL_REQUEST_TEMPLATE.md) — filling in Description, Related Issue, Type of Change, Testing, and the Checklist keeps reviews consistent. (`Fixes #1942` is already there — nice.)
- In that checklist, please tick the CI checks that apply to this change and leave the rest as N/A. This diff only touches a Python script in the `mural` skill, so the markdown / docs / PowerShell checks don't apply; the relevant ones are skill structure validation (`npm run validate:skills`), Python lint (`npm run lint:py`), and the skill's own pytest suite, plus **Tests added** and **Documentation updated** under Required Checks (see inline comments).

I've left a few inline comments — one test gap and a docstring update I'd like to see before merge, plus a couple of optional / consistency notes. Thanks again for contributing!

File: .github/skills/experimental/mural/scripts/mural/_backends.py (hunk 2 above)
Comment (akazlow) [reply .github/skills/experimental/mural/scripts/mural/_backends.py:239]: fixed!

File: .github/skills/experimental/mural/scripts/mural/_backends.py (hunk 4 above)
Comment (akazlow) [reply .github/skills/experimental/mural/scripts/mural/_backends.py]: fixed

**Comment by WilliamBerryiii** [comment]
Thanks for this fix — the fallback itself is solid. It only fires when keyring holds zero credentials, and it enforces the same 0600/owner contract (`enforce_file_perms=True`) as explicit `file` mode, so nothing secure is shadowed.

One follow-on I'd like to fold in while we're in this code, because I don't want us to settle for living on the plaintext file when a working keyring is right there: **when keyring is available-but-empty and the file is populated, let's promote the credentials into the keyring** so the operator actually ends up on the secure store.

Important on *where*: please don't put the promotion inside `resolve_backend` — that's the read path (`auth status`, `list`, etc.), and a silent keyring write there can trigger an OS unlock prompt or fail on a locked vault during what should be a passive read. Instead, do the promotion in the auth/write flows (`auth login` / `auth bootstrap`) that already expect to touch credentials: detect "keyring available + empty + file populated," write the creds into keyring, **read them back to confirm the write succeeded**, and only then clear the file. If the keyring write fails, keep the file and fall back exactly like today. The empty-keyring fallback in `resolve_backend` can stay as-is for the immediate read, plus a one-line hint that promotion will happen on next login.

Net effect: this PR''s fix keeps auth working immediately, and the next real auth touch moves people off plaintext and onto the keyring — without making passive reads do surprising writes.

To make this easy to pick up, here''s a task-researcher prompt you can run to scope the change:

<details>
<summary>Task-researcher prompt</summary>

```text
Research and produce an implementation plan for adding "file -> keyring credential promotion"
to the Mural CLI credential subsystem.

Context / goal:
When MURAL_CREDENTIAL_BACKEND=auto and the OS keyring backend is available but holds no Mural
credentials, while the per-user credential file IS populated, the CLI should promote (copy) the
file credentials into the keyring so the operator ends up on the more secure store instead of
living on the plaintext file indefinitely. PR #2135 already fixed the immediate bug (empty keyring
no longer shadows usable file creds); this task is the follow-on promotion behavior.

Hard constraint on placement:
Do NOT perform the keyring write inside resolve_backend(). That function is the read path called by
passive commands (auth status, list, etc.); a silent keyring write there can trigger an OS unlock
prompt or fail on a locked vault. The promotion MUST live in the auth write flows
(_cmd_auth_login and/or _cmd_auth_bootstrap) where touching credentials is already expected.

Files to investigate (all under .github/skills/experimental/mural/scripts/mural/):
- _backends.py: resolve_backend(), KeyringBackend, FileBackend, _emit, the _KeyringUnavailable
  sentinel, and the one-shot warn state via _state.seen_fallback_warn().
- _credentials.py: _backend_has_credentials(), _check_credential_file_perms(), _service_name_for(),
  _resolve_credential_file(), and the _KNOWN_CREDENTIAL_KEYS / _PROFILE_REQUIRED_KEYS constants.
- _cli_auth.py: _cmd_auth_bootstrap() and the auth login command — these are the intended homes for
  promotion. Note bootstrap already persists via resolve_backend(profile).
- _constants.py: credential key names and env var constants.
- tests/test_credential_storage.py: existing backend-resolution and fallback tests to mirror.

Required behavior to design and validate:
1. Promotion trigger: auto mode, keyring backend constructs successfully, keyring has none of
   _KNOWN_CREDENTIAL_KEYS for the profile service, and the file backend is populated (reuse
   _backend_has_credentials with enforce_file_perms=True so the 0600/owner contract still gates it).
2. Promotion action: write each populated credential key from the file into the KeyringBackend for
   _service_name_for(profile).
3. Verify-before-delete: after writing, read the values back from keyring and confirm they match the
   file values. Only clear/remove the credential file AFTER a verified successful write. If the file
   is left, document why.
4. Failure handling: if the keyring write or read-back fails (locked vault, _KeyringUnavailable, any
   keyring error), keep the file untouched and fall back to file exactly like today. Never lose
   credentials.
5. Non-interactive safety: respect ENV_NONINTERACTIVE / CI — do not trigger OS prompts in
   non-interactive contexts; decide whether promotion is skipped or attempted silently there.
6. Idempotency + one-shot logging: promotion runs once per profile per process; emit a clear WARNING/
   INFO describing the promotion and the file cleanup, deduped via the existing _state seen-set pattern.
7. Windows: _check_credential_file_perms is a POSIX-only no-op; confirm promotion still behaves
   correctly on Windows and note the residual ACL gap.

Deliverables:
- A short evidence-backed findings doc (current resolution/bootstrap flow, where promotion fits).
- A concrete implementation plan listing exact functions to add/modify and their call sites.
- A test plan extending test_credential_storage.py: keyring-empty+file-populated promotion success,
  keyring-write-failure preserves file + falls back, verify-before-delete, non-interactive skip,
  one-shot dedupe, and a Windows-path case.
Do not write production code in this pass — produce research + plan only.
```

</details>

Happy to pair on the command surface if useful.

**Comment by akazlow** [comment]
> Thanks for this fix — the fallback itself is solid. It only fires when keyring holds zero credentials, and it enforces the same 0600/owner contract (`enforce_file_perms=True`) as explicit `file` mode, so nothing secure is shadowed.
> 
> One follow-on I'd like to fold in while we're in this code, because I don't want us to settle for living on the plaintext file when a working keyring is right there: **when keyring is available-but-empty and the file is populated, let's promote the credentials into the keyring** so the operator actually ends up on the secure store.
> 
> Important on _where_: please don't put the promotion inside `resolve_backend` — that's the read path (`auth status`, `list`, etc.), and a silent keyring write there can trigger an OS unlock prompt or fail on a locked vault during what should be a passive read. Instead, do the promotion in the auth/write flows (`auth login` / `auth bootstrap`) that already expect to touch credentials: detect "keyring available + empty + file populated," write the creds into keyring, **read them back to confirm the write succeeded**, and only then clear the file. If the keyring write fails, keep the file and fall back exactly like today. The empty-keyring fallback in `resolve_backend` can stay as-is for the immediate read, plus a one-line hint that promotion will happen on next login.
> 
> Net effect: this PR''s fix keeps auth working immediately, and the next real auth touch moves people off plaintext and onto the keyring — without making passive reads do surprising writes.
> 
> To make this easy to pick up, here''s a task-researcher prompt you can run to scope the change:
> 
> Task-researcher prompt
> ```
> Research and produce an implementation plan for adding "file -> keyring credential promotion"
> to the Mural CLI credential subsystem.
> 
> Context / goal:
> When MURAL_CREDENTIAL_BACKEND=auto and the OS keyring backend is available but holds no Mural
> credentials, while the per-user credential file IS populated, the CLI should promote (copy) the
> file credentials into the keyring so the operator ends up on the more secure store instead of
> living on the plaintext file indefinitely. PR #2135 already fixed the immediate bug (empty keyring
> no longer shadows usable file creds); this task is the follow-on promotion behavior.
> 
> Hard constraint on placement:
> Do NOT perform the keyring write inside resolve_backend(). That function is the read path called by
> passive commands (auth status, list, etc.); a silent keyring write there can trigger an OS unlock
> prompt or fail on a locked vault. The promotion MUST live in the auth write flows
> (_cmd_auth_login and/or _cmd_auth_bootstrap) where touching credentials is already expected.
> 
> Files to investigate (all under .github/skills/experimental/mural/scripts/mural/):
> - _backends.py: resolve_backend(), KeyringBackend, FileBackend, _emit, the _KeyringUnavailable
>   sentinel, and the one-shot warn state via _state.seen_fallback_warn().
> - _credentials.py: _backend_has_credentials(), _check_credential_file_perms(), _service_name_for(),
>   _resolve_credential_file(), and the _KNOWN_CREDENTIAL_KEYS / _PROFILE_REQUIRED_KEYS constants.
> - _cli_auth.py: _cmd_auth_bootstrap() and the auth login command — these are the intended homes for
>   promotion. Note bootstrap already persists via resolve_backend(profile).
> - _constants.py: credential key names and env var constants.
> - tests/test_credential_storage.py: existing backend-resolution and fallback tests to mirror.
> 
> Required behavior to design and validate:
> 1. Promotion trigger: auto mode, keyring backend constructs successfully, keyring has none of
>    _KNOWN_CREDENTIAL_KEYS for the profile service, and the file backend is populated (reuse
>    _backend_has_credentials with enforce_file_perms=True so the 0600/owner contract still gates it).
> 2. Promotion action: write each populated credential key from the file into the KeyringBackend for
>    _service_name_for(profile).
> 3. Verify-before-delete: after writing, read the values back from keyring and confirm they match the
>    file values. Only clear/remove the credential file AFTER a verified successful write. If the file
>    is left, document why.
> 4. Failure handling: if the keyring write or read-back fails (locked vault, _KeyringUnavailable, any
>    keyring error), keep the file untouched and fall back to file exactly like today. Never lose
>    credentials.
> 5. Non-interactive safety: respect ENV_NONINTERACTIVE / CI — do not trigger OS prompts in
>    non-interactive contexts; decide whether promotion is skipped or attempted silently there.
> 6. Idempotency + one-shot logging: promotion runs once per profile per process; emit a clear WARNING/
>    INFO describing the promotion and the file cleanup, deduped via the existing _state seen-set pattern.
> 7. Windows: _check_credential_file_perms is a POSIX-only no-op; confirm promotion still behaves
>    correctly on Windows and note the residual ACL gap.
> 
> Deliverables:
> - A short evidence-backed findings doc (current resolution/bootstrap flow, where promotion fits).
> - A concrete implementation plan listing exact functions to add/modify and their call sites.
> - A test plan extending test_credential_storage.py: keyring-empty+file-populated promotion success,
>   keyring-write-failure preserves file + falls back, verify-before-delete, non-interactive skip,
>   one-shot dedupe, and a Windows-path case.
> Do not write production code in this pass — produce research + plan only.
> ```
> 
> Happy to pair on the command surface if useful.

Great idea. I'll get started on that and send you a message

File: .github/skills/experimental/mural/tests/test_credential_storage.py
Comment (rezatnoMsirhC) [inline .github/skills/experimental/mural/tests/test_credential_storage.py:1]: **Missing companion test: empty-keyring + bad file permissions**

The PR adds `test_auto_keyring_unavailable_enforces_file_permissions` to verify that the unavailable-keyring path enforces file permissions (raising `MuralError` on bad perms). The empty-keyring fallback path in `_backends.py` also calls `_backend_has_credentials` with `enforce_file_perms=True`, so it can raise `MuralError` the same way. There is no companion test for that path.

A test mirroring the unavailable-keyring version would close the gap:

```python
@pytest.mark.skipif(os.name == "nt", reason="POSIX-only credential file permission semantics")
def test_auto_empty_keyring_enforces_file_permissions(
    self,
    mural_module: Any,
    monkeypatch: pytest.MonkeyPatch,
    tmp_path: pathlib.Path,
) -> None:
    _isolate_credential_env(monkeypatch, tmp_path)
    monkeypatch.setenv("MURAL_KEYRING_BACKEND", "keyrings.alt.file.PlaintextKeyring")

    checked_paths: list[pathlib.Path] = []

    def _raise_bad_perms(path: pathlib.Path, environ: dict[str, str]) -> None:
        checked_paths.append(path)
        raise mural_module.MuralError("simulated bad permissions")

    monkeypatch.setattr(mural_module, "_check_credential_file_perms", _raise_bad_perms)

    file_path = mural_module._resolve_credential_file("default", os.environ)
    file_path.parent.mkdir(parents=True, exist_ok=True)
    file_path.write_text("MURAL_CLIENT_ID=from-file\n", encoding="utf-8")

    with pytest.raises(mural_module.MuralError, match="simulated bad permissions"):
        mural_module.resolve_backend("default")

    assert checked_paths == [file_path]
```

**Review by rezatnoMsirhC — DISMISSED** [review]

File: .github/skills/experimental/mural/tests/test_credential_storage.py
Comment (akazlow) [reply .github/skills/experimental/mural/tests/test_credential_storage.py:1]: Implemented. I am double checking work between meetings then will push up tomorrow.

File: .github/skills/experimental/mural/scripts/mural/_cli_auth.py [hunk 6]
```diff
@@ -152,6 +155,119 @@ def _save_token_store_locked(path: pathlib.Path, data: dict[str, Any]) -> None:
...
+            keyring_backend.set(service, key, value)
+            roundtrip = keyring_backend.get(service, key)
+        except _KeyringUnavailable as exc:
+            _emit(
+                f"file-to-keyring credential promotion failed for profile "
+                f"{profile_name!r}; keeping file copy at {file_path} "
+                f"({exc})",
+                level=logging.WARNING,
+            )
+            return
```
Comment (rezatnoMsirhC) [inline .github/skills/experimental/mural/scripts/mural/_cli_auth.py:247]: **Blocking: roll back partial keyring writes**

If an earlier credential writes successfully and a later `set()` or round-trip verification fails, this return leaves the earlier value in keyring. On the next process, `resolve_backend()` considers keyring populated when any known key exists, so it selects the incomplete keyring instead of the intact file. I reproduced this by allowing `MURAL_CLIENT_ID` to write and failing `MURAL_CLIENT_SECRET`; auto mode then returned a keyring with no client secret.

Please track the keys written by this promotion attempt and delete them before returning from any later write or verification failure. A regression test should fail the second write and confirm that auto mode still resolves to the complete file backend.

**Review by rezatnoMsirhC — COMMENTED** [review]
Rereview comments on the credential-promotion update.

File: .github/skills/experimental/mural/tests/test_credential_storage.py [hunk 7]
```diff
@@ -429,6 +545,140 @@ def _raise_unavailable(self: Any) -> None:
...
+class TestCredentialPromotion:
+    def test_promotes_file_credentials_to_keyring_and_removes_file(
+        self,
+        mural_module: Any,
+        monkeypatch: pytest.MonkeyPatch,
+        tmp_path: pathlib.Path,
+        caplog: pytest.LogCaptureFixture,
+    ) -> None:
+        mural_module._state.seen_fallback_warn().clear()
+        _isolate_credential_env(monkeypatch, tmp_path)
```
Comment (rezatnoMsirhC) [inline .github/skills/experimental/mural/tests/test_credential_storage.py:613]: **Blocking: isolate this promotion test from `CI=true`**

GitHub Actions sets `CI=true`, which makes `_is_noninteractive_context()` skip promotion. These positive promotion tests do not clear that variable, so the required Mural Python test job currently fails before exercising the behavior it intends to verify.

Please clear `CI` in `_isolate_credential_env()` or explicitly in the positive promotion tests. The noninteractive test can remain deterministic by setting `MURAL_NONINTERACTIVE=1` itself.

**Review by WilliamBerryiii — COMMENTED** [review]
Review discussion is being reconciled against the updated branch head. The addressed threads will be resolved individually; final approval will follow the new head CI result.

File: .github/skills/experimental/mural/scripts/mural/_backends.py (hunk 1 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/mural/scripts/mural/_backends.py:241]: System.Collections.Hashtable.Body

File: .github/skills/experimental/mural/scripts/mural/_backends.py (hunk 2 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/mural/scripts/mural/_backends.py:239]: System.Collections.Hashtable.Body

File: .github/skills/experimental/mural/scripts/mural/_backends.py (hunk 3 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/mural/scripts/mural/_backends.py]: System.Collections.Hashtable.Body

File: .github/skills/experimental/mural/scripts/mural/_backends.py (hunk 4 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/mural/scripts/mural/_backends.py]: System.Collections.Hashtable.Body

File: .github/skills/experimental/mural/scripts/mural/_backends.py (hunk 5 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/mural/scripts/mural/_backends.py]: System.Collections.Hashtable.Body

File: .github/skills/experimental/mural/tests/test_credential_storage.py
Comment (WilliamBerryiii) [reply .github/skills/experimental/mural/tests/test_credential_storage.py:1]: System.Collections.Hashtable.Body

File: .github/skills/experimental/mural/scripts/mural/_cli_auth.py (hunk 6 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/mural/scripts/mural/_cli_auth.py:247]: System.Collections.Hashtable.Body

File: .github/skills/experimental/mural/tests/test_credential_storage.py (hunk 7 above)
Comment (WilliamBerryiii) [reply .github/skills/experimental/mural/tests/test_credential_storage.py:691]: System.Collections.Hashtable.Body

**Review by WilliamBerryiii — APPROVED** [review]
Validated the updated head after merging current main: 113 checks passed, 5 expected checks skipped, and no checks failed or were cancelled. The credential fallback and promotion failure-path coverage are addressed, and all review threads are resolved.

**Review by agreaves-ms — APPROVED** [review]
